Add a `bookkeeper` plugin

niftynei opened the pull request:
Bookkeeper plugin keeps track of all the coin movements that happen on the attached clightning node. It also exposes a number of commands for inspecting these events, as well as some 'on top' accounting to figure out what you've spent on onchain fees.

- `bkpr-listbalances`: shows all accounts and their balances (a channel is an account)
- `bkpr-listaccountevents`: dump of all raw events (incl onchain fee updates, as they happen). Can be filtered by `account`
- `bkpr-inspect`: show the onchain footprint of this `account`. Nice for seeing unilateral close resolution progress.
- `bkpr-listincome`: list every 'income impacting' event (internal transfers etc are filtered). Defaults to consolidating fees.
- `bkpr-dumpincomecsv`: Dump out the income events to a csv formatted for a given provider. Options include: cointracker, koinly, harmony, and quickbooks.
- `bkpr-channelsapy`: Summary of fees earned on incoming/outgoing payments and fee leases on a channel by channel basis.

 
Note that any coin movements that occurred before the bookkeeper plugin was initialized won't be counted.

rustyrussell commented:
Ok, I've merged the prerequisites....

cdecker commented on plugins/bkpr/bookkeeper.c:
```diff
@@ -73,12 +74,297 @@ static struct command_result *json_list_balances(struct command *cmd,
+				    currency,
+				    NULL, 0,
+				    timestamp);
+	db_begin_transaction(db);
+	log_channel_event(db, acct, chan_ev);
+	db_commit_transaction(db);
```
General: it's often easier and even faster to just wrap the entire event handling in a single large DB transaction rather than trying to do fine-grained control, which may lead to accidentally nested transactions (-> abort), and many transactions being started and committed. Also the one-to-one mapping between event and DB transaction ensure that we don't get partially handled events in case of a crash.

cdecker commented on plugins/bkpr/bookkeeper.c:
```diff
@@ -25,6 +27,158 @@ static struct db *db ;
+		struct channel_event *chan;
+		struct chain_event *chain;
+		struct onchain_fee *fee;
+		u64 lowest = 0;
```
Why not set this to UINT64_MAX? Allows you to treat the "no lowest" case identical to "higher lowest" case.

In addition is seems that this assumes that there'll always be a fee to add at the end, wouldn't it be easier to do a merge-sort like structure where each iteration pops the lowest and prints that before restarting the iteration? This code might get the ordering wrong if we have two consecutive events of the same type followed by a later event, since this will at most print one event for each type before switching to the next type (though there may be a structure in the model I'm not considering here).

niftynei replied:
> this will at most print one event for each type before switching to the next type

We only print one item every loop, there are `continue`s at each print action that ensures that there's only one per.

> this assumes that there'll always be a fee to add at the end

Since we continue after print, you'll never reach the fee print line if you're out of fees -- another item will have printed and the loop continued.

Generally speaking, the routine here is:
- elect the top item of each list
- from all (still populated) elected items, which has the lowest timestamp?
- print that (with some preference for channel, then chain, then fees in case of a tie)
- return to election phase

I could merge the lowest timestamp identification and 'topmost item' selection.

cdecker reviewed (commented):
Just some preliminary remarks and questions up to commit bd3252a3876f1e45cd4f51ffe78b18b1f0e063ce, will continue the review later today.

niftynei commented:
Based on comments from @cdecker, I rewrote the way that the `account_get_balance` function works to use the `SUM()` methods in SQL. Perhaps foolishly, I embraced the 'maybe this account has multiple currencies' idea and rewrote the interface to return the balances segmented by 'currency', which had some rippling impacts throughout the rest of the stuff, but it's mostly been folded in. I'm a glutton for big patchsets aren't I.

https://github.com/ElementsProject/lightning/pull/5071/commits/1f364f9ae8937e95abed39b48b1f9e8c013f06af#diff-ae7e6b3b74048f72d90eac244f2b7eda49ee6637cf323de3f3cdd2b9af179ed8R197

 I wasn't sure about the best way to handle adding a prefix to a table name for the tables, since the shared database code expects certain tables. A better refactorer would probably pull the part of the database migrations out that are expected/core into a top-level thing or something..

michaelfolkson commented:
Concept ACK

Is the plan that you'll get this working in this repo first and then move the plugin specific stuff (bookkeeper database etc) to the plugins [repo](https://github.com/lightningd/plugins)? This is essentially laying the groundwork for others to write accounting plugins longer term right? (If there was reason to of course, perhaps this bookkeeper plugin will literally do everything a user could wish for re accounting).

rustyrussell commented:
Looks like valgrind found something...

bertmiller commented:
lgtm!

rustyrussell commented:
After discussion with @niftynei I am bumping this to next release....

cdecker commented on plugins/bkpr/recorder.c:
```diff
@@ -0,0 +1,545 @@
+	struct db_stmt *stmt;
+	struct chain_event *e;
+
+	if (spending_txid) {
```
I see, that might require a bit more normalization of the conditions, and become longer, but it should work. It's a tradeoff though, and if it gets too long and wonky it might definitely not be worth doing :+1:

ZmnSCPxj commented on plugins/bkpr/db.c:
```diff
@@ -0,0 +1,168 @@
+ * It is used to migrate existing databases from a prevoius state, based on
+ * string indicies */
+static struct migration db_migrations[] = {
+	{SQL("CREATE TABLE version (version INTEGER);"), NULL},
```
Probably a tall ask, but why a separate database?  Could we use `datastore` instead?

niftynei replied:
iiuc `datastore` is designed for key-value storage; this application requires more complex relational data storage.

ZmnSCPxj replied:
All relational data storage is based on key-value ultimately (tables have a row ID per row that is the key to the tuple, and indices map tuple values to row ID keys), though I do understand wanting not to have to reimplement SQL when a perfectly good library already exists.

In particular, this seems to be recordkeeping that is "safe" to lose; you lose your records, but none of your funds are at risk if the `bookkeeper`-specific database is lost due to inevitable hardware failure.  But I think you should also point this out in the docs, then, that since it is not stored in `lightningd.sqlite3` or in the PostgreSQL  database, then any automated backup/replication setup the user uses does ***not*** apply to the `bookkeeper` data.

rustyrussell commented on connectd/test/run-netaddress.c:
```diff
@@ -233,4 +233,5 @@ int main(int argc, char *argv[])
 	assert(IsIPv6(&wa));
 
 	common_shutdown();
+	return 0;
 }
```
This is unnecessary in modern C.  And I guess from the commit message, this is simply a rind left over from a previous commit which added common_shutdown() and is now redundant?

niftynei replied:
removed! (deleted whole commit, was extraneous)

rustyrussell commented on plugins/bkpr/account.c:
```diff
@@ -1,17 +1,18 @@
 
 struct account *new_account(const tal_t *ctx,
-			    const char *name STEALS,
+			    const char *name,
```
Technically, this is now TAKES (since tal_strdup() can take).

niftynei replied:
i just removed the thing b/c we don't ever call this with `take`... (and it does strdup)

rustyrussell replied:
But it *would* take, because tal_strdup() does this...

rustyrussell commented on plugins/bkpr/bookkeeper.c:
```diff
@@ -1791,6 +1800,11 @@ int main(int argc, char *argv[])
+				  "string",
+				  "Location of the bookkeeper database",
+				  charp_option, &db_dsn),
 		    NULL);
```
bookkeeper options should be documented in doc/lightningd-config.5.md, too.

rustyrussell replied:
Oh, they are!  Great!

rustyrussell commented on doc/lightningd-config.5.md:
```diff
@@ -221,6 +221,13 @@ automatically by `lightningd`.
 Directory to keep the accounts.sqlite3 database file in.
 Defaults to lightning-dir.
 
+ **bookkeeper-db**=*DSN*
```
Add `[plugin ``bookkeeper``]` to the end of this (like bcli options do)?

rustyrussell reviewed (approved):
Only reviewed half, but minor comments.  Mainly, needs rebase and thwacking CI AFAICT...

rustyrussell commented:
Rebased on master, and reordered and fixed deps so it's possible to bisect, also fixed db_fatal() in plugin to show up in logs.  Did some minor fixups which were obvious.

The following tests are broken:
* tests/test_closing.py::test_onchain_their_unilateral_out
* tests/test_closing.py::test_penalty_htlc_tx_timeout
* tests/test_opening.py::test_zeroconf_forward 
* tests/test_opening.py::test_zeroconf_open

The zeroconf ones seem like a conflict introduced in the rebase:

```
lightningd-2 2022-07-19T07:23:59.871Z **BROKEN** plugin-bookkeeper: Received channel event, but no account exists 6cb8d54790e10d5e368681f33a15e9b1a651ed218996cfb804cc5825e4d010c1
```

rustyrussell commented:
Pushed a patch to try to catch that crash in CI...

rustyrussell commented:
OK, some memleak reports need plugging:

```
lightningd-1 2022-07-23T19:11:11.803Z **BROKEN** plugin-bookkeeper: MEMLEAK: 0x1b32448
lightningd-1 2022-07-23T19:11:11.851Z **BROKEN** plugin-bookkeeper:   label=plugins/bkpr/recorder.c:1048:struct account
lightningd-1 2022-07-23T19:11:11.851Z **BROKEN** plugin-bookkeeper:   backtrace:
lightningd-1 2022-07-23T19:11:11.852Z **BROKEN** plugin-bookkeeper:     ccan/ccan/tal/tal.c:442 (tal_alloc_)
lightningd-1 2022-07-23T19:11:11.853Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/recorder.c:1048 (stmt2account)
lightningd-1 2022-07-23T19:11:11.853Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/recorder.c:1109 (find_account)
lightningd-1 2022-07-23T19:11:11.853Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1606 (parse_and_log_channel_move)
lightningd-1 2022-07-23T19:11:11.853Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1713 (json_coin_moved)
lightningd-1 2022-07-23T19:11:11.853Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1421 (ld_command_handle)
lightningd-1 2022-07-23T19:11:11.853Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1491 (ld_read_json_one)
lightningd-1 2022-07-23T19:11:11.853Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1511 (ld_read_json)
lightningd-1 2022-07-23T19:11:11.854Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:59 (next_plan)
lightningd-1 2022-07-23T19:11:11.854Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:407 (do_plan)
lightningd-1 2022-07-23T19:11:11.854Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:417 (io_ready)
lightningd-1 2022-07-23T19:11:11.854Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/poll.c:453 (io_loop)
lightningd-1 2022-07-23T19:11:11.854Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1708 (plugin_main)
lightningd-1 2022-07-23T19:11:11.854Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1812 (main)
lightningd-1 2022-07-23T19:11:11.875Z **BROKEN** plugin-bookkeeper:     ../csu/libc-start.c:308 (__libc_start_main)
lightningd-1 2022-07-23T19:11:11.875Z **BROKEN** plugin-bookkeeper:   parents:
lightningd-1 2022-07-23T19:11:11.875Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1378:struct command
```

This needs to be freed before calling another request, which I think is fine.

```
lightningd-1 2022-07-23T19:11:11.879Z **BROKEN** plugin-bookkeeper: MEMLEAK: 0x1b5d0b8
lightningd-1 2022-07-23T19:11:11.879Z **BROKEN** plugin-bookkeeper:   label=plugins/bkpr/bookkeeper.c:1568:struct channel_event
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:   backtrace:
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:     ccan/ccan/tal/tal.c:442 (tal_alloc_)
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1568 (parse_and_log_channel_move)
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1713 (json_coin_moved)
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1421 (ld_command_handle)
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1491 (ld_read_json_one)
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1511 (ld_read_json)
lightningd-1 2022-07-23T19:11:11.880Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:59 (next_plan)
lightningd-1 2022-07-23T19:11:11.881Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:407 (do_plan)
lightningd-1 2022-07-23T19:11:11.881Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:417 (io_ready)
lightningd-1 2022-07-23T19:11:11.881Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/poll.c:453 (io_loop)
lightningd-1 2022-07-23T19:11:11.881Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1708 (plugin_main)
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1812 (main)
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper:     ../csu/libc-start.c:308 (__libc_start_main)
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper:   parents:
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1378:struct command
```

This may need a notleak(), or you can free `e`.

```
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper: MEMLEAK: 0x1b5b268
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper:   label=plugins/bkpr/bookkeeper.c:1643:enum mvt_tag[]
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper:   backtrace:
lightningd-1 2022-07-23T19:11:11.882Z **BROKEN** plugin-bookkeeper:     ccan/ccan/tal/tal.c:442 (tal_alloc_)
lightningd-1 2022-07-23T19:11:11.883Z **BROKEN** plugin-bookkeeper:     ccan/ccan/tal/tal.c:471 (tal_alloc_arr_)
lightningd-1 2022-07-23T19:11:11.883Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1643 (parse_tags)
lightningd-1 2022-07-23T19:11:11.883Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1686 (json_coin_moved)
lightningd-1 2022-07-23T19:11:11.883Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1421 (ld_command_handle)
lightningd-1 2022-07-23T19:11:11.883Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1491 (ld_read_json_one)
lightningd-1 2022-07-23T19:11:11.895Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1511 (ld_read_json)
lightningd-1 2022-07-23T19:11:11.895Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:59 (next_plan)
lightningd-1 2022-07-23T19:11:11.895Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:407 (do_plan)
lightningd-1 2022-07-23T19:11:11.895Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/io.c:417 (io_ready)
lightningd-1 2022-07-23T19:11:11.896Z **BROKEN** plugin-bookkeeper:     ccan/ccan/io/poll.c:453 (io_loop)
lightningd-1 2022-07-23T19:11:11.896Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1708 (plugin_main)
lightningd-1 2022-07-23T19:11:11.896Z **BROKEN** plugin-bookkeeper:     plugins/bkpr/bookkeeper.c:1812 (main)
lightningd-1 2022-07-23T19:11:11.897Z **BROKEN** plugin-bookkeeper:     ../csu/libc-start.c:308 (__libc_start_main)
lightningd-1 2022-07-23T19:11:11.897Z **BROKEN** plugin-bookkeeper:   parents:
lightningd-1 2022-07-23T19:11:11.897Z **BROKEN** plugin-bookkeeper:     plugins/libplugin.c:1378:struct command
```

This one is a bit trickier.  You might be able to allocate tags off tmpctx?  Actually, same with the others probably...

niftynei commented:
flake is #5408

rustyrussell commented on plugins/bkpr/incomestmt.c:
```diff
@@ -105,6 +105,30 @@ static struct income_event *onchainfee_to_income(const tal_t *ctx,
+	esc = json_escape(ctx, dupe);
+	safe_str = tal_strcat(ctx, "\"", esc->s);
+	safe_str = tal_strcat(ctx, safe_str, "\"");
+	return safe_str;
```
This function produces a lot of allocations off ctx, when most are transient.

rustyrussell replied:
This screams `tal_fmt(ctx, "\"%s\"", esc->s)`?

rustyrussell commented on plugins/bkpr/incomestmt.c:
```diff
@@ -105,6 +105,30 @@ static struct income_event *onchainfee_to_income(const tal_t *ctx,
+ * by wrapping the desc in double-quotes ("). But what if
+ * there's already double-quotes? Well we swap these to
+ * single-quotes (') and then use the json_escape function */
+static char *csv_safe_str(const tal_t *ctx, char *input)
```
const char *input TAKES?

rustyrussell commented on plugins/bkpr/incomestmt.c:
```diff
@@ -105,6 +105,30 @@ static struct income_event *onchainfee_to_income(const tal_t *ctx,
+
+	/* Update the double-quotes in place */
+	dupe = tal_strdup(ctx, input);
+	while (dupe[i] != '\0') {
```
Ugliest for loop ever...

niftynei replied:
Updated to actually be for loop.

rustyrussell commented on plugins/bkpr/bookkeeper.c:
```diff
@@ -553,7 +553,6 @@ static bool new_missed_channel_account(struct command *cmd,
 				       u64 timestamp)
 {
-	struct chain_event *chain_ev;
 	size_t i, j;
```
Not a big fan of this commit. Seems like you need separate "channel proposed" and "channel confirmed" events, and you're shortcutting?

niftynei replied:
I agree. @ddustin heard me complaining about this last week... should have just done it then TM.

niftynei replied:
Ok, fixed up so that the event log always rolls forward.

rustyrussell reviewed (commented):
Not sure about violating the append-only rule here for zeroconf.

And some minor cleanup which can be done post

niftynei commented:
https://github.com/ElementsProject/lightning/pull/5071/commits/da742b54f9e9762040a7b8dfc1f802db8a6bd0d9: now emits a new event type "channel_proposed" for zeroconfs

https://github.com/ElementsProject/lightning/pull/5071/commits/4822c5490c5237e4011ae7f6a2d1d016307a6ad2: keeps all the events. `inspect` now does a little sleigh of hand now as it hides the `proposed` event after an `open` happens.

rustyrussell commented on plugins/bkpr/recorder.c:
```diff
@@ -225,7 +225,8 @@ static struct chain_event **find_txos_for_tx(const tal_t *ctx,
 				     "  e.utxo_txid"
 				     ", e.outnum"
-				     ", e.spending_txid NULLS FIRST"));
+				     ", e.spending_txid NULLS FIRST"
```
NULLS FIRST, hmm, I've never seen that before! TIL...

rustyrussell reviewed (approved):
Ack da5d7022a7f020fc8a58867c1137030355098bd3

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/pull/5071/commits/45229517a251c66bb148f2a585677c934e1e3815
